### DISCUSSION TITLE
Prevent focus returning to trigger element after closing

### DISCUSSION THREAD

**Discussion by BjornDCode** [tailwindlabs/headlessui#708] [Ideas]
The Dialog, Popover and Menu components currently return focus to the 'trigger' element after closing. For most use cases this is great, but it would be nice to be able to override this behaviour. 

We use a lot of Headless UI elements to open menus with actions that add new items to a list or similar behaviour. In those cases, it would be nice to not have focus return to the trigger element so we can manually focus the new element added to the list.

According to the WAI-ARIA spec it's valid to not return focus to the trigger element if makes more sense to move focus to a newly added element.

> When a dialog closes, focus returns to the element that invoked the dialog unless either:
The invoking element no longer exists. Then, focus is set on another element that provides logical work flow.
The work flow design includes the following conditions that can occasionally make focusing a different element a more logical choice:
It is very unlikely users need to immediately re-invoke the dialog.
The task completed in the dialog is directly related to a subsequent step in the work flow.
For example, a grid has an associated toolbar with a button for adding rows. the Add Rows button opens a dialog that prompts for the number of rows. After the dialog closes, focus is placed in the first cell of the first new row.

Source: https://www.w3.org/TR/wai-aria-practices/#dialog_modal

**Comment by roydigerhund** [comment]
Did you find a workaround for this? I'm using the `Menu` to add custom filters and would like to focus the input field of the filter after clicking on a `Menu.Item`. But the `Menu.Button` steals the focus here. 

My current solution is to wait a 50ms before focusing the new input element but this is not an elegant solution.

**Reply by BjornDCode** [reply]
We ended up using the Dropdown Menu from Radix (https://www.radix-ui.com/docs/primitives/components/dropdown-menu). It has a `onCloseAutoFocus` callback prop where you can call `event.preventDefault`

**Comment by mseele** [comment]
I want to prevent focus returning to trigger element after closing for the headlessui Vue Dialog, too.

@RobinMalfait Is there any Feature (via a property) planned to support this or do you suggest any workaround for this?

We need to move away from the headlessui Vue Dialog (which we use because we use a payed tailwindui template) in this situation if there is no solution for this problem. And having to wait for 50ms and re-trigger focus is not an option...

**Reply by prashank25** [reply]
For anyone looking, I managed to solve this issue by disabling the trigger button when the modal is open.

```
<button
  @click="modalOpen = true"
  :disabled="modalOpen"
>
  Open Modal
</button>
```

**Reply by jerrebm** [reply]
@Prashank25 This approach worked, but somehow breaks the open transition animation of the modal. 

I'm now calling `document.activeElement.blur()` on the trigger element before setting the open state to true, which seems to do the trick without breaking animations.

**Comment by haykkh** [comment]
We have also run into this issue on our end with Listboxes. Our solution for right now is to watch the `open` state  using a wrapper component that takes the slot prop and emits changes back up to our parent component, then watching the new/old value of that and listening for the focus event on the trigger to return focus back to our element:

```vue
<script setup>
import { Listbox, ListboxButton, ListboxOptions, ListboxOption } from '@headlessui/vue'
import { useEventListener } from '@vueuse/core'
import WrapperComponent from '...'


const props = defineProps<{
  returnFocusElement: HTMLElement | null
}>()

const buttonRef = ref<HTMLElement | null>(null)

const openClose = ({ newOpen, oldOpen }) => {
  if (oldOpen === true && newOpen === false) {
    useEventListener(buttonRef, 'focus' () => props.returnFocusElement.focus())
  }
}
</script>

<template>
  <Listbox v-slot="{ open }">
    <WrapperComponent :open="open" @wrapper-update="openClose">
      <ListboxButton ref="buttonref" />

      ...ListboxOptions etc
    </WrapperComponent>
  </Listbox>
</template>
```

This seems very inelegant and error prone, with lots of things to "keep track of". Especially when it seems like a simple boolean flag check before focusing the button would solve this (look for `buttonRef)?.focus()` in [listbox.ts](https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-vue/src/components/listbox/listbox.ts)). 

Maybe someone who understands the library a bit better than me can comment on how simple that last bit actually would be 😬

**Comment by vince-roy** [comment]
The blurring solution (https://github.com/tailwindlabs/headlessui/discussions/708#discussioncomment-4714800) works in @headlessui/react@v1.6.4 but not in the latest @headlessui/react@v1.7.x.
